File: meshers/gmsh/mesh_data.cpp
#include <gmsh.h>
#include <iostream>
#include <fstream>
#include <vector>
#include <set>
#include <array>
#include <algorithm>
#include <string>
#include <map>

using Line2 = std::array<int, 2>;
using Line3 = std::array<int, 3>;

std::string format_nodes(const std::vector<double>& coords) {
    std::string out;
    size_t n = coords.size() / 3;
    for (size_t i = 0; i < n; ++i) {
        out += std::to_string(coords[3*i+0]) + " " +
               std::to_string(coords[3*i+1]) + " " +
               std::to_string(coords[3*i+2]) + "  ";
    }
    return out;
}

int main(int argc, char** argv) {
    if (argc < 3) {
        std::cerr << "need mesh hash and dir in the command line" << std::endl;
        return 1;
    }

    std::string mesh_file = std::string(argv[2]) + "/" + argv[1] + ".msh";
    std::ifstream infile(mesh_file);
    if (!infile.good()) {
        std::cerr << "mesh file does not exist" << std::endl;
        return 1;
    }
    infile.close();

    std::cout << "1" << std::endl;
    gmsh::initialize();
    gmsh::option::setNumber("General.Terminal", 0);
    gmsh::open(mesh_file);

    std::map<std::string, std::string> mesh;

    // Nodes
    std::vector<std::size_t> nodeTags;
    std::vector<double> nodeCoords, parametricCoords;
    gmsh::model::mesh::getNodes(nodeTags, nodeCoords, parametricCoords);

    mesh["nodes"] = format_nodes(nodeCoords);

    // Surface edges (lines)
    std::set<Line2> lines2;
    std::set<Line3> lines3;
    std::string surfaces_edges_set;

    std::vector<int> elementTypes;
    std::vector<std::vector<std::size_t>> elementTags, elementNodeTags;

    // 2 = surface elements
    gmsh::model::mesh::getElements(elementTypes, elementTags, elementNodeTags, 2);

    for (size_t i = 0; i < elementTypes.size(); ++i) {
        int type = elementTypes[i];
        const auto& nodes = elementNodeTags[i];
        // 3-node triangles (type==2) or possibly 6-node (type==9)
        size_t j = 0;
        if (type == 2) {
            // Linear triangle: 3 nodes per element
            for (size_t e = 0; e < elementTags[i].size(); ++e) {
                std::array<int,2> l1 = {static_cast<int>(nodes[j+0])-1, static_cast<int>(nodes[j+1])-1};
                std::array<int,2> l2 = {static_cast<int>(nodes[j+1])-1, static_cast<int>(nodes[j+2])-1};
                std::array<int,2> l3 = {static_cast<int>(nodes[j+2])-1, static_cast<int>(nodes[j+0])-1};
                std::sort(l1.begin(), l1.end());
                std::sort(l2.begin(), l2.end());
                std::sort(l3.begin(), l3.end());
                lines2.insert(l1);
                lines2.insert(l2);
                lines2.insert(l3);
                j += 3;
            }
        } else if (type == 9) {
            // Quadratic triangle: 6 nodes per element
            for (size_t e = 0; e < elementTags[i].size(); ++e) {
                std::array<int,3> l1 = {static_cast<int>(nodes[j+0])-1, static_cast<int>(nodes[j+3])-1, static_cast<int>(nodes[j+1])-1};
                std::array<int,3> l2 = {static_cast<int>(nodes[j+1])-1, static_cast<int>(nodes[j+4])-1, static_cast<int>(nodes[j+2])-1};
                std::array<int,3> l3 = {static_cast<int>(nodes[j+2])-1, static_cast<int>(nodes[j+5])-1, static_cast<int>(nodes[j+0])-1};
                std::sort(l1.begin(), l1.end());
                std::sort(l2.begin(), l2.end());
                std::sort(l3.begin(), l3.end());
                lines3.insert(l1);
                lines3.insert(l2);
                lines3.insert(l3);
                j += 6;
            }
        }
    }

    for (const auto& l : lines2) {
        surfaces_edges_set += std::to_string(l[0]) + " " + std::to_string(l[1]) + " -1 ";
    }
    for (const auto& l : lines3) {
        surfaces_edges_set += std::to_string(l[0]) + " " + std::to_string(l[1]) + " " + std::to_string(l[2]) + " -1 ";
    }
    mesh["surfaces_edges_set"] = surfaces_edges_set;

    std::cout << "2" << std::endl;

    // Surface faces, one per each physical group
    std::map<int, std::string> surfaces_faces_set;
    std::vector<std::pair<int,int>> physicals;
    gmsh::model::getPhysicalGroups(physicals);

    for (const auto& physical : physicals) {
        int dim = physical.first;
        int physical_tag = physical.second;
        if (dim == 2) {
            std::string faces;
            std::vector<int> entities;
            gmsh::model::getEntitiesForPhysicalGroup(dim, physical_tag, entities);
            for (int entity : entities) {
                std::vector<int> types;
                std::vector<std::vector<std::size_t>> tags, nodetags;
                gmsh::model::mesh::getElements(types, tags, nodetags, dim, entity);
                for (size_t i = 0; i < types.size(); ++i) {
                    int type = types[i];
                    for (size_t j = 0; j < tags[i].size(); ++j) {
                        if (type == 2) {
                            // 3-node triangle
                            faces += std::to_string(static_cast<int>(nodetags[i][j*3+0])-1) + " " +
                                     std::to_string(static_cast<int>(nodetags[i][j*3+1])-1) + " " +
                                     std::to_string(static_cast<int>(nodetags[i][j*3+2])-1) + " ";
                        } else if (type == 9) {
                            // 6-node triangle
                            faces += std::to_string(static_cast<int>(nodetags[i][j*6+0])-1) + " " +
                                     std::to_string(static_cast<int>(nodetags[i][j*6+3])-1) + " " +
                                     std::to_string(static_cast<int>(nodetags[i][j*6+5])-1) + " ";
                            faces += std::to_string(static_cast<int>(nodetags[i][j*6+1])-1) + " " +
                                     std::to_string(static_cast<int>(nodetags[i][j*6+4])-1) + " " +
                                     std::to_string(static_cast<int>(nodetags[i][j*6+3])-1) + " ";
                            faces += std::to_string(static_cast<int>(nodetags[i][j*6+2])-1) + " " +
                                     std::to_string(static_cast<int>(nodetags[i][j*6+5])-1) + " " +
                                     std::to_string(static_cast<int>(nodetags[i][j*6+4])-1) + " ";
                            faces += std::to_string(static_cast<int>(nodetags[i][j*6+3])-1) + " " +
                                     std::to_string(static_cast<int>(nodetags[i][j*6+4])-1) + " " +
                                     std::to_string(static_cast<int>(nodetags[i][j*6+5])-1) + " ";
                        }
                    }
                }
            }
            surfaces_faces_set[physical_tag] = faces;
        }
    }

    mesh["surfaces_faces_set"] = "";
    for (const auto& kv : surfaces_faces_set) {
        mesh["surfaces_faces_set"] += "\"" + std::to_string(kv.first) + "\": \"" + kv.second + "\", ";
    }

    std::cout << "3" << std::endl;

    gmsh::finalize();

    // Output as JSON (simple, not using a JSON lib)
    std::ofstream out(std::string(argv[2]) + "/" + argv[1] + "-data.json");
    out << "{\n";
    out << "\"nodes\": \"" << mesh["nodes"] << "\",\n";
    out << "\"surfaces_edges_set\": \"" << mesh["surfaces_edges_set"] << "\",\n";
    out << "\"surfaces_faces_set\": {" << mesh["surfaces_faces_set"] << "}\n";
    out << "}" << std::endl;
    out.close();

    return 0;
}

File: meshers/gmsh/mesh_data.py
#!/usr/bin/python3
import sys
sys.path.append("../../../../bin")
import gmsh
import os
import json

# Add a segment, always in increasing order, to make duplicate detection easy
def addLine2(lines, lines_set, tags, first, second):
  a = int(tags[first])
  b = int(tags[second])
  entry = (min(a-1, b-1), max(a-1, b-1))
  if entry not in lines_set:
    lines.append([entry[0], entry[1]])
    lines_set.add(entry)
  return

def addLine3(lines, lines_set, tags, first, second, third):
  a = int(tags[first])
  b = int(tags[second])
  c = int(tags[third])
  # Always store in the lowest, middle, largest order (for consistency)
  sorted_entry = tuple(sorted([a-1, b-1, c-1]))
  if sorted_entry not in lines_set:
    lines.append(list(sorted_entry))
    lines_set.add(sorted_entry)
  return

# ------------------------------------------

if (len(sys.argv) < 3):
  print("need mesh hash and dir in the command line")
  sys.exit(1)

mesh_file = "%s/%s.msh" % (sys.argv[2], sys.argv[1])
if not os.path.exists(mesh_file):
  print("mesh file does not exist")
  sys.exit(1)

print("1", flush=True)
gmsh.initialize()
gmsh.option.setNumber("General.Terminal", 0)
gmsh.open(mesh_file)

# TODO: read whether the mesh is curved or not
curved = 0

mesh = {}

# nodes ------------
mesh["nodes"] = ""
tags, coord, _ = gmsh.model.mesh.getNodes()
maxtag = max(tags)
for i in range(maxtag):
  mesh["nodes"] += "{:6g} {:6g} {:6g}  ".format(coord[3*i+0], coord[3*i+1], coord[3*i+2])

# surface edges ------------
mesh["surfaces_edges_set"] = ""
elements = gmsh.model.mesh.getElements(2)
n_elements = len(elements)
lines = []
lines_set = set()
i = 0
k = 0
for type in elements[0]:
  j = 0
  if type == 2 or (curved == 0 and type == 9):    # 3-node triangle
    for element in elements[1][i]:
      addLine2(lines, lines_set, elements[2][i], j+0, j+1)
      addLine2(lines, lines_set, elements[2][i], j+1, j+2)
      addLine2(lines, lines_set, elements[2][i], j+2, j+0)
      j += 3 if type == 2 else 6 
      k += 1
  elif type == 9:  # 6-node triangle
    for element in elements[1][i]:
      addLine3(lines, lines_set, elements[2][i], j+0, j+3, j+1)
      addLine3(lines, lines_set, elements[2][i], j+1, j+4, j+2)
      addLine3(lines, lines_set, elements[2][i], j+2, j+5, j+0)
      j += 6
      k += 1
  i += 1

print("2", flush=True)

n_lines = len(lines)
k = 0
for line in lines:
  if (len(line) == 2):
    mesh["surfaces_edges_set"] += "{:d} {:d} -1 ".format(line[0], line[1])
  elif (len(line) == 3):
    mesh["surfaces_edges_set"] += "{:d} {:d} {:d} -1 ".format(line[0], line[1], line[2])
  k += 1  
print("3", flush=True)

# surface faces, one per each physical group ------------
mesh["surfaces_faces_set"] = {}
physicals = gmsh.model.getPhysicalGroups()
k = 0
for physical in physicals:
  dim = physical[0]
  physical_tag = physical[1]
  if (dim == 2):
    mesh["surfaces_faces_set"][physical_tag] = ""
    for entity in gmsh.model.getEntitiesForPhysicalGroup(dim, physical_tag):
      types, tags, nodetags = gmsh.model.mesh.getElements(dim, entity)
      for i in range(len(types)):
        for j in range(len(tags[i])):
          if types[i] == 2 or (curved == 0 and types[i] == 9):
            N = 6 if (types[i] == 9) else 3
            # for triangles remove the last int and the -1
            mesh["surfaces_faces_set"][physical_tag] += "{:d} {:d} {:d} ".format(int(nodetags[i][j*N+0])-1, int(nodetags[i][j*N+1])-1, int(nodetags[i][j*N+2])-1)
          elif types[i] == 9:
            N = 6
            mesh["surfaces_faces_set"][physical_tag] += "{:d} {:d} {:d} ".format(int(nodetags[i][j*N+0])-1, int(nodetags[i][j*N+3])-1, int(nodetags[i][j*N+5])-1)
            mesh["surfaces_faces_set"][physical_tag] += "{:d} {:d} {:d} ".format(int(nodetags[i][j*N+1])-1, int(nodetags[i][j*N+4])-1, int(nodetags[i][j*N+3])-1)
            mesh["surfaces_faces_set"][physical_tag] += "{:d} {:d} {:d} ".format(int(nodetags[i][j*N+2])-1, int(nodetags[i][j*N+5])-1, int(nodetags[i][j*N+4])-1)
            mesh["surfaces_faces_set"][physical_tag] += "{:d} {:d} {:d} ".format(int(nodetags[i][j*N+3])-1, int(nodetags[i][j*N+4])-1, int(nodetags[i][j*N+5])-1)
          k += 1  

print("4", flush=True)
gmsh.finalize()

with open("%s/%s-data.json" % (sys.argv[2], sys.argv[1]), "w", encoding ='utf8') as json_file:
  json.dump(mesh, json_file)
